### PR TITLE
Proposal/MCMS Management Library

### PR DESCRIPTION
**Pull request by akhilchainani**
### What

Including a set of Go tools/lib function to use MCMS + Timelock Contracts

### CODE REVIEW

**Comment by 0xAustinWang** [comment]
I need to fork off this for my work, but let's follow more idiomatic golang practices such as using `/pkg` or `/internal` directories.

**Comment by akhilchainani** [comment]
> I need to fork off this for my work, but what do you think of following more idiomatic golang practices such as using `/pkg` or `/internal` directories.

Yeah no objections from me - @cmalec how about you?

File: tools/configwrappers/config.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		Quorum:       quorum,
+		Signers:      signers,
+		GroupSigners: groupSigners,
+	}
+}
+
+func NewConfigFromRaw(rawConfig gethwrappers.ManyChainMultiSigConfig) *Config {
+	groupToSigners := make([][]common.Address, len(rawConfig.GroupQuorums))
+	for _, signer := range rawConfig.Signers {
+		groupToSigners[signer.Group] = append(groupToSigners[signer.Group], signer.Addr)
```
Comment (cmalec) [inline tools/configwrappers/config.go:32]: is it safe to assume the `signer.Group` is within the `len(rawConfig.GroupQuorums)`?

File: tools/configwrappers/config.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	groups := make([]Config, len(rawConfig.GroupQuorums))
+	for i, quorum := range rawConfig.GroupQuorums {
+		signers := groupToSigners[i]
```
Comment (cmalec) [inline tools/configwrappers/config.go:37]: is `group` just an index? since above we set `groupToSigners[signer.Group]` but now we are just doing `groupToSigners[i]`

File: tools/configwrappers/config.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+		}
+	}
+
+	for i, parent := range rawConfig.GroupParents {
+		if i > 0 && groups[i].Quorum > 0 {
+			groups[parent].GroupSigners = append(groups[parent].GroupSigners, groups[i])
+		}
+	}
+
+	return &groups[0]
```
Comment (cmalec) [inline tools/configwrappers/config.go:55]: why return only a pointer to the first element after all this? maybe i dont have enough context into the contract

File: tools/executable/executable.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+				ChainIdentifier: t.ChainIdentifier,
+				Parameter:       "chain metadata",
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *ExecutableMCMSProposal) ToExecutor(clients map[string]ContractDeployBackend) (*Executor, error) {
```
Comment (connorwstein) [inline tools/executable/executable.go:78]: This factory sort of pattern would allow for a caller to create multiple executors for the same proposal which we would not want right? I think it makes more sense to have a executor object instantiated with clients which has a function like 
```
func (e Executor) ExecuteProposal(ep ExecutableMCMSProposal) error {}
```

File: tools/managed/mcm_with_timelock.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		return &errors.ErrInvalidMinDelay{
+			ReceivedMinDelay: m.MinDelay,
+		}
+	}
+
+	return nil
+}
+
+func (m *MCMSWithTimelockProposal) ToExecutableMCMSProposal() (executable.ExecutableMCMSProposal, error) {
```
Comment (connorwstein) [inline tools/managed/mcm_with_timelock.go:72]: This laying of embedding is pretty hard to follow and I could imagine it leading to fields empty unintentionally/difficult initialization. Can we invert the design a bit and just have constructors for the 2 cases we have here? Example
```
package mcms

import (
	"crypto"
	"encoding/binary"
	"fmt"
	"math/big"
)

type TimelockBatch struct {
	Ops []Op `json:"ops"`
	Description string `json:"description"`
}

type Op struct {
	ChainIdentifer string         `json:"chain_identifer,omitempty"`
	To             common.Address `json:"to,omitempty"`
	Value          big.Int        `json:"value"`
	Data           []byte         `json:"data,omitempty"`
}

type Proposal struct {
	Version              string      `json:"version"`
	ValidUntil           uint32      `json:"validUntil"`
	Signatures           []Signature `json:"signatures"`
	OverridePreviousRoot bool        `json:"overridePreviousRoot"`

	// Map of chain identifier to chain metadata
	ChainMetadata map[string]ExecutableMCMSChainMetadata `json:"chainMetadata"`
	Operations []Op `json:"ops"`
}


func (p Proposal) Sign(signer func(h [32]byte) (Signature, error), clients map[string]bind.ContractBackend) error {
	root, err := generateRoot(clients, p)
	if err != nil {
		return err
	}
	var validUntilBytes [32]byte
	binary.BigEndian.PutUint32(validUntilBytes[28:], p.ValidUntil) // Place the uint32 in the last 4 bytes
	hashToSign := crypto.Keccak256Hash(root, validUntilBytes[:])
	sig, err := signer(toEthSignedMessageHash(hashToSign))
	if err != nil {
		return err
	}
	p.Signatures = append(p.Signatures, sig)
	return nil	
}

type ChainMetadata struct {
	NonceOffset uint64         `json:"nonceOffset" :"nonce_offset"`
	MCMAddress  common.Address `json:"mcmAddress" :"mcm_address"`
}

func NewProposal(
	version string, // Type?
	validUntil uint32, // Type?
	overridePreviousRoot bool,
	chainMetadata map[string]ExecutableMCMSChainMetadata,
	operations []Op,
) (Proposal, error) {
	// Validate parameters
	return Proposal{
	}, nil
}

func NewProposalFromTimelockBatches(
	batches []TimelockBatch,
	version string, 
	validUntil uint32, // Type?
	overridePreviousRoot bool,
	chainMetadata map[string]ExecutableMCMSChainMetadata,
	operations []Op,
	) (Proposal, error) {
	// Validate parameters
	return Proposal{
		
	}, nil
}
```
I'm not sure we need the executor abstraction either with above approach

**Review by connorwstein — CHANGES_REQUESTED** [review]

File: src/ManyChainMultiSig.sol [hunk 6]
```diff
@@ -398,11 +398,11 @@ contract ManyChainMultiSig is Ownable2Step {
         bool clearRoot
     ) external onlyOwner {
         if (signerAddresses.length == 0 || signerAddresses.length > MAX_NUM_SIGNERS) {
-            revert OutOfBoundsNumOfSigners();
+            revert("OutOfBoundsNumOfSigners()");
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply src/ManyChainMultiSig.sol:401]: lol oops, I was doing some testing and wasn't able to get the revert code, this shouldn't be here

File: go.mod [hunk 7]
```diff
@@ -2,34 +2,95 @@ module github.com/smartcontractkit/ccip-owner-contracts
 
 go 1.22
 
-require github.com/ethereum/go-ethereum v1.13.8
+require github.com/ethereum/go-ethereum v1.14.7
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply go.mod]: Hm is 1.13.10 okay? Seems as though the new simulated backend is available starting from 1.13.10, so won't have to rewrite a bunch of the tests if we can use 1.13.10

File: tools/configwrappers/config.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	"math/big"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/errors"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/gethwrappers"
+)
+
+// Config is a struct that holds all the configuration for the owner contracts
+type Config struct {
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/configwrappers/config.go:13]: The existing config is really difficult to interpret - I have a guess that if we don't include something like this, consumers will build their own more intuitive structures for managing the membership of the MCMS config

File: tools/executable/executor.go [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/gethwrappers"
+)
+
+type Executor struct {
+	Proposal         *ExecutableMCMSProposal
+	Tree             *MerkleTree
+	RootMetadatas    map[string]gethwrappers.ManyChainMultiSigRootMetadata
+	Operations       map[string][]gethwrappers.ManyChainMultiSigOp
+	ChainAgnosticOps []gethwrappers.ManyChainMultiSigOp
+	Caller           *Caller
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/executable/executor.go:22]: yeah you're probably right, was trying to not bloat one struct with too much responsibility

File: tools/executable/merkle_tree.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
+package executable
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+type MerkleTree struct {
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/executable/merkle_tree.go:8]: yeah its new, we could use another solution out there, but I couldn't find one that was very clearly "Industry-standard" implementation, but this is easy to swap out if we wanna do that later

File: tools/executable/merkle_tree.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	// Root hash of the merkle tree
+	Root common.Hash
+
+	// Layers of the merkle tree, starting from the leaves
+	Layers [][]common.Hash
+}
+
+func NewMerkleTree(leaves []common.Hash) *MerkleTree {
+	tree := &MerkleTree{
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/executable/merkle_tree.go:17]: whats the correct way to implement something like this?

File: tools/executable/executable.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+	if len(m.ChainMetadata) == 0 {
+		return &errors.ErrNoChainMetadata{}
+	}
+
+	return nil
+}
+
+type ExecutableMCMSProposal struct {
+	ExecutableMCMSProposalBase
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/executable/executable.go]: Its just some shared attributes that are used in other structs as well, so it allows us to reuse the same `Validate` function but its not required, just introduces a bit of redundancy but probably at the expense of readability

File: tools/executable/executable.go (hunk 4 above)
Comment (akhilchainani) [reply tools/executable/executable.go:66]: Executor isn't just used for execution though, it also holds the fully constructed MerkleTree which is used for signing as well though, so having a single `Execute` function wouldn't suffice. Executor just represents the Full expanded Proposal with all machine executable details required for signing, validating, and executing

File: tools/managed/mcm_with_timelock.go (hunk 5 above)
Comment (akhilchainani) [reply tools/managed/mcm_with_timelock.go:72]: Hm let me think about this for a bit, but in the meantime, How would we construct a proposal from a file in this case then? Because signatures will be getting aggregated through some file, so we need to be able to unmarshall the JSON into a given proposal structure

File: go.mod (hunk 7 above)
[the remark this answers is not in the document]
Comment (connorwstein) [reply go.mod]: no the sim backend breaking change is the one that breaks core

File: tools/configwrappers/wrapper.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+type WrappedManyChainMultisig struct {
+	gethwrappers.ManyChainMultiSig
+}
+
+type ContractDeployBackend interface {
+	bind.ContractBackend
+	bind.DeployBackend
+}
+
+func DeployAndConfigureManyChainMultisig(auth *bind.TransactOpts, backend ContractDeployBackend, config *Config) (common.Address, []*types.Transaction, *WrappedManyChainMultisig, error) {
```
[the remark this answers is not in the document]
Comment (connorwstein) [reply tools/configwrappers/wrapper.go]: I could see the MCMS Config struct potentially being helpful, but just a Config.ToArgs() function would be sufficient for use at a higher level. The thin wrapper methods on a WrappedManyChainMultisig functions don't provide much value

File: tools/managed/mcm_with_timelock.go (hunk 5 above)
Comment (akhilchainani) [reply tools/managed/mcm_with_timelock.go:72]: > This laying of embedding is pretty hard to follow and I could imagine it leading to fields empty unintentionally/difficult initialization.
Not following why this is the case - validations ensure all required fields exist, and this is the only entrypoint that consumers need to concern themselves with.

File: tools/configwrappers/wrapper.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/configwrappers/wrapper.go]: Doesn't it become additional and unnecessary cognitive load to know that during SetRoot, signatures need to be sorted by their dervied address in increasing order. Because thats a requirement for SetRoot calls to succeed

[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/configwrappers/wrapper.go]: Hm ok yeah thats fair

File: tools/executable/merkle_tree.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (connorwstein) [reply tools/executable/merkle_tree.go:17]:
```
func NewMerkleTree(leaves ..) *MerkleTree {
	layers := make(..)
	for ... {
		// build layers
	}
	// We apply the reference at call frame boundary
	return &MerkleTree{
		layers: layers,
	}
}
```

File: tools/executable/utils.go [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+type Operation struct {
+	To    common.Address
+	Data  string
+	Value uint64
+}
+
+type ChainOperation struct {
+	ChainIdentifier string
+	Operation
+}
```
Comment (carte7000) [inline tools/executable/utils.go:24]: Should we have json annotation on this struct like on `ExecutableMCMSProposal`

```suggestion
type Operation struct {
	To    common.Address `json:"to"`
	Data  string `json:"data"`
	Value uint64 `json:"value"`
}

type ChainOperation struct {
	ChainIdentifier string `json:"chainId"`
	Operation
}
```

File: tools/errors/errors.go [hunk 15]
```diff
@@ -0,0 +1,148 @@
+package errors
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// ErrInvalidChainID is the error for an invalid chain ID.
+type ErrInvalidChainID struct {
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/errors/errors.go:10]: We don't get the typing of certain errors doing it this way then right:?

File: tools/executable/utils.go (hunk 14 above)
Comment (akhilchainani) [reply tools/executable/utils.go:24]: Executable should in theory never be directly written/read from JSON, so I lean no, but it also doesn't hurt.

File: tools/timelock_proposal/operations.go [hunk 16]
```diff
@@ -0,0 +1,8 @@
+package timelock_proposal
```
Comment (0xAustinWang) [inline tools/timelock_proposal/operations.go:1]: https://go.dev/blog/package-names

> They are lower case, with no under_scores or mixedCaps. 

We can just call it timelock

File: tools/mcms_proposal/executor.go [hunk 17]
```diff
@@ -0,0 +1,282 @@
+package mcms_proposal
```
Comment (0xAustinWang) [inline tools/mcms_proposal/executor.go:1]: https://go.dev/blog/package-names

> They are lower case, with no under_scores or mixedCaps. 

We can just call it mcmsproposal

File: tools/timelock_proposal/mcm_with_timelock.go [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/errors"
+	owner "github.com/smartcontractkit/ccip-owner-contracts/tools/gethwrappers"
+	"github.com/smartcontractkit/ccip-owner-contracts/tools/mcms_proposal"
+)
+
+var ZERO_HASH = common.BytesToHash([]byte{})
```
Comment (0xAustinWang) [inline tools/timelock_proposal/mcm_with_timelock.go:15]: https://github.com/ethereum/go-ethereum/blob/v1.14.8/core/types/block.go#L484

Zero hash can just be `common.Hash{}`

File: tools/timelock_proposal/mcm_with_timelock.go [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+)
+
+type MCMSWithTimelockProposal struct {
+	mcms_proposal.Proposal
+
+	Operation TimelockOperation `json:"operation"` // Always 'schedule', 'cancel', or 'bypass'
+
+	// TODO: this should be configurable as a human-readable string
+	// i.e. 1d, 1w, 1m, 1y
+	MinDelay string `json:"minDelay"`
```
Comment (0xAustinWang) [inline tools/timelock_proposal/mcm_with_timelock.go:37]: why human readable instead of a `time.Duration`?

File: tools/timelock_proposal/mcm_with_timelock.go [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+	Transactions []BatchChainOperation `json:"transactions"`
+}
+
+func (m *MCMSWithTimelockProposal) Validate() error {
+	if err := m.Proposal.Validate(); err != nil {
+		return err
+	}
+
+	switch m.Operation {
+	case Schedule, Cancel, Bypass:
```
Comment (0xAustinWang) [inline tools/timelock_proposal/mcm_with_timelock.go:52]: Do we need this validation? The proposal should fail to Unmarshal because it's a defined type `TimelockOperation`

File: tools/timelock_proposal/mcm_with_timelock.go [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+	mcmOnly.ChainMetadata = make(map[string]mcms_proposal.ChainMetadata)
+	for chain, metadata := range m.ChainMetadata {
+		mcmOnly.ChainMetadata[chain] = mcms_proposal.ChainMetadata{
+			NonceOffset: metadata.NonceOffset,
+			MCMAddress:  metadata.MCMAddress,
+		}
+	}
+
+	// Convert transactions into timelock wrapped transactions
+	for _, t := range m.Transactions {
```
Comment (0xAustinWang) [inline tools/timelock_proposal/mcm_with_timelock.go:89]: We can make this a separate function for better unit testing

File: tools/signing/sign_plain_key.go [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	// Sign the payload
+	unmarshalledSig := mcms_proposal.Signature{}
+	err = json.Unmarshal(sig, &unmarshalledSig)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Add signature to proposal
```
Comment (0xAustinWang) [inline tools/signing/sign_plain_key.go:56]: nit: could insert it sorted

File: tools/timelock_proposal/mcm_with_timelock.go (hunk 21 above)
Comment (akhilchainani) [reply tools/timelock_proposal/mcm_with_timelock.go:89]: yeah need to add more extensive unit tests for this regardless but yeah sure

File: tools/timelock_proposal/mcm_with_timelock.go (hunk 20 above)
Comment (akhilchainani) [reply tools/timelock_proposal/mcm_with_timelock.go:52]: will it definitely fail to unmarshal? either way, will add a test for it and if thats the case then I'll remove this validation

**Comment by akhilchainani** [comment]
> post-convo with Christian, we should have separate release cycles and versioning for the proposal-generation side, and the proposal-execution side. If the way to achieve this is using a different repository, we should do so.
> 
> The goal here is that proposal-generation would be consumed from the chainlink repo, and execution would be consumed from `chainlink-deployments` repo. Decoupling the two makes it so we can pin dependencies against specific versions without running into the issue of breaking upgrades

I think seperate repos is gratuitous, we can define multiple go packages in this repo itself and also have them release independently as well if need be

File: tools/mcms_proposal/encoding_test.go [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+			PostOpCount:          big.NewInt(1),
+			OverridePreviousRoot: false,
+		},
+	}
+
+	tree, err := buildMerkleTree(chainIdentifiers, rootMetadatas, ops)
+	assert.NoError(t, err)
+	assert.NotNil(t, tree)
+	assert.NotEmpty(t, tree.Root)
+}
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/mcms_proposal/encoding_test.go:210]: we already have this test across both the `merkle_tree` and `executor` tests, both of them contstruct tree and validates them and another tries to execute the proposal too so the tree is inherently constructed correctly

File: tools/mcms_proposal/encoding.go [hunk 24]
```diff
@@ -0,0 +1,159 @@
+package mcms_proposal
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/mcms_proposal/encoding.go:1]: Hm potentially, I think all of the components that are chain-specific should eventually move out into their own chain family specific folders. Lets keep thinking about this for now because of the top of my head:
* simulation
* execution
* operation/metadata encoding 
 
are all going to be chain specific so we should think more about how we want to structure it. I don't think we need to do this in this PR, we can address it in a follow up PR

File: tools/configwrappers/config.go [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+			}
+		}
+
+		if !found {
+			return false
+		}
+	}
+
+	// Compare all group signers in second exist in first (order doesn't matter)
+	for i := range other.GroupSigners {
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/configwrappers/config.go:140]: yeah that works too

File: tools/configwrappers/config.go [hunk 26]
```diff
@@ -0,0 +1,245 @@
...
+
+	return gethwrappers.ManyChainMultiSigConfig{
+		GroupQuorums: groupQuorums,
+		GroupParents: groupParents,
+		Signers:      signers,
+	}
+}
+
+func (c *Config) Equals(other *Config) bool {
+	if c.Quorum != other.Quorum {
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/configwrappers/config.go:107]: wanted to account for the case where configurations have the same signer sets/group sets but are presented in a different order

File: go.mod [hunk 27]
```diff
@@ -2,34 +2,96 @@ module github.com/smartcontractkit/ccip-owner-contracts
...
+	golang.org/x/sys v0.19.0 // indirect
+	golang.org/x/text v0.14.0 // indirect
+	golang.org/x/time v0.3.0 // indirect
+	google.golang.org/protobuf v1.27.1 // indirect
+	gopkg.in/natefinch/lumberjack.v2 v2.0.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
 	rsc.io/tmplfunc v0.0.3 // indirect
 )
+
+replace github.com/ethereum/go-ethereum v1.13.10 => github.com/ethereum/go-ethereum v1.13.10
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply go.mod:97]: it supposedly pins the version to that particular minor version, becasue every time I was doing running `go mod tidy` it was upgrade geth

File: .github/workflows/go-test.yml [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+      run: |
+        echo "Uncommitted changes found after 'go mod tidy'. Please commit changes."
+        exit 1
+      if: ${{ failure() }}
+
+    - name: Build the project
+      run: go build ./...
+
+    - name: Run tests
+      run: go test ./...
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply .github/workflows/go-test.yml:48]: yeah we can scrap one of them

File: tools/timelock_proposal/operations.go (hunk 16 above)
Comment (akhilchainani) [reply tools/timelock_proposal/operations.go:1]: gonna say timelock proposal because it doesn't have ALL of the timelock logic here (i.e. geth wrappers is in the getwrappers package)

File: tools/mcms_proposal/encoding.go [hunk 29]
```diff
@@ -0,0 +1,159 @@
...
+		if !ok {
+			return nil, &errors.ErrMissingChainDetails{
+				ChainIdentifier: chainID,
+				Parameter:       "current op count",
+			}
+		}
+
+		currentTxCount, ok := txCounts[chainID]
+		if !ok {
+			return nil, &errors.ErrMissingChainDetails{
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/mcms_proposal/encoding.go]: I don't follow what you mean by this, can you elaborate?

File: tools/signing/sign_plain_key.go (hunk 22 above)
Comment (akhilchainani) [reply tools/signing/sign_plain_key.go:56]: don't really see a significant advantage to doing this, seems liike more code for what would a not noticeable difference in performance at the scale we expect this array to be at

File: tools/mcms_proposal/encoding.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (0xAustinWang) [reply tools/mcms_proposal/encoding.go]: Maybe this is my lack of golang experience, I was thinking if there was a way to return the error object directly instead of the address of the error. 

However, the way you have it for custom errors seems idioimatic: https://gobyexample.com/custom-errors

File: tools/timelock_proposal/operations.go (hunk 16 above)
Comment (0xAustinWang) [reply tools/timelock_proposal/operations.go]: I was thinking our path could be more descriptive and look something like:

import(
   "github.com/smartcontractkit/ccip-owner-contracts/pkg/proposals/mcms"
   "github.com/smartcontractkit/ccip-owner-contracts/pkg/proposals/timelock"
)

That way the information about what the package does is described in the import path, while still sticking to golang conventions

File: tools/timelock_proposal/mcm_with_timelock.go (hunk 19 above)
Comment (0xAustinWang) [reply tools/timelock_proposal/mcm_with_timelock.go]: Aren't we assuming that these proposals will be generated by a chainlink-deployments migration? I would assume that the input into the migration would be human readable. 

I would argue that since we're building a library, it's ok to assume users wrapper will be able to generate a `time.Duration`

File: tools/signing/sign_plain_key.go (hunk 22 above)
Comment (0xAustinWang) [reply tools/signing/sign_plain_key.go:56]: The value isn't for performance, but for deterministic ordering when we process the proposal afterwards. We either insert the signatures sorted, or we sort the signatures later when executing / checking roots. 

Just a tiny detail, but it feels cleaner to me

Comment (akhilchainani) [reply tools/signing/sign_plain_key.go:56]: > but for deterministic ordering when we process the proposal afterwards. We either insert the signatures sorted, or we sort the signatures later when executing / checking roots.

This is only true if we can guarantee that signatures are only ever inserted in this way, which we can't always guarantee because after signing, library users can theoretically reorder the sigs array. IMO, We should only transform when we need to transform to ensure that we can guarantee that when we try to SetRoot we don't get any failures

File: tools/timelock_proposal/mcm_with_timelock.go (hunk 19 above)
Comment (akhilchainani) [reply tools/timelock_proposal/mcm_with_timelock.go]: right but this proposal format will be written somewhere as sigs are aggregated, so this value will need to get written somewhere, what would that look like?

File: tools/timelock_proposal/operations.go (hunk 16 above)
Comment (akhilchainani) [reply tools/timelock_proposal/operations.go]: right but this wouldn't be accurate because both of the MCMS and Timelock geth wrappers are in an independent directory. We can restructure this in a future iteration, maybe

Comment (0xAustinWang) [reply tools/timelock_proposal/operations.go]: The import path *is* the description of this library. The path suggested above describes that this is logic specific to timelock proposals, and mcms proposals.

File: tools/mcms_proposal/encoding.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/mcms_proposal/encoding.go]: yeah I think this is just the way they are instantiated, because `error` is still the return time

File: tools/timelock_proposal/mcm_with_timelock.go (hunk 19 above)
Comment (0xAustinWang) [reply tools/timelock_proposal/mcm_with_timelock.go]: Seems fine to leave it as a string, if we use the https://pkg.go.dev/time#ParseDuration function from the time library. 

In that case, would it make sense to use the errors directly from the library rather than wrapping them? Is there additional information that we add from wrapping it compared to https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/time/format.go;l=1686?

File: tools/mcms_proposal/utils.go [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+
+func mapMCMAddresses(metadatas map[string]ChainMetadata) map[string]common.Address {
+	m := make(map[string]common.Address)
+	for k, v := range metadatas {
+		m[k] = v.MCMAddress
+	}
+	return m
+}
+
+func mapSignatures(signatures []Signature) []gethwrappers.ManyChainMultiSigSignature {
```
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/mcms_proposal/utils.go]: yeah no preference on the name really, I'll change it to one of the suggestions you had. Not sure about sorting here, because I'm not sure if this function will be used elsewhere other than SetRoot. If its only used in SetRoot, it doesn't matter, but if it is gonna be used elsewhere, I wanna avoid adding in the complexity of sorting in places where its not relevant to the library consumers. sorting is only relevant when broadcasting onchain so I think we should confine that logic to that function

File: tools/signing/sign_plain_key.go (hunk 22 above)
Comment (akhilchainani) [reply tools/signing/sign_plain_key.go:56]: This is also related to another comment above but I'll reiterate here: 

>  because I'm not sure if this function will be used elsewhere other than SetRoot. If its only used in SetRoot, it doesn't matter, but if it is gonna be used elsewhere, I wanna avoid adding in the complexity of sorting in places where its not relevant to the library consumers. sorting is only relevant when broadcasting onchain so I think we should confine that logic to that function

File: tools/timelock_proposal/mcm_with_timelock.go (hunk 19 above)
Comment (akhilchainani) [reply tools/timelock_proposal/mcm_with_timelock.go]: yeah probably don't need the custom error type here, will remove

File: tools/timelock_proposal/mcm_with_timelock.go (hunk 20 above)
Comment (akhilchainani) [reply tools/timelock_proposal/mcm_with_timelock.go]: Hm, following up on this, atleast in the unit tests, I can pass in an arbitrary strings and it doesn't seem to error unless I call validate. Maybe json.Unmarshal will catch it if that happens but some of these will be programatically constructed so probably wanna catch this error sooner than when reading from file

File: tools/timelock_proposal/operations.go (hunk 16 above)
Comment (akhilchainani) [reply tools/timelock_proposal/operations.go]: ohhhh I missed the `proposal` part, yeah this structure works for me

File: tools/errors/errors.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (akhilchainani) [reply tools/errors/errors.go:10]: kk sounds good, gonna fix this in a follow up PR to not block other work that depends on this lib

**Review by carte7000 — APPROVED** [review]

**Review by 0xAustinWang — APPROVED** [review]
